### PR TITLE
Organization API

### PR DESCRIPTION
**Pull request by sleshchenko**
### What does this PR do?
Adds Organization API. It includes:
- Adds organizational account to make able to link workspace and organization
  - Organizational account allow to make organization name unique in common scope with user name
- Adds ability for user to create and manage organizations
- Adds ability for user to manage members of organizations by permissions service

### What issues does this PR fix or reference?
https://github.com/codenvy/codenvy/issues/552

### Tests written?
Yes

### CODE REVIEW

**Comment by sleshchenko** [comment]
@akorneta @evoevodin @skabashnyuk @mshaposhnik Please review my changes

File: wsmaster/codenvy-hosted-api-organization/src/main/java/com/codenvy/organization/api/OrganizationManager.java [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+
+    @Inject
+    public OrganizationManager(OrganizationDao organizationDao,
+                               MemberDao memberDao,
+                               EventService eventService,
+                               @Named("che.account.reserved_names") String[] reservedNames) {
+        this.organizationDao = organizationDao;
+        this.memberDao = memberDao;
+        this.eventService = eventService;
+        this.reservedNames = Sets.newHashSet(reservedNames);
```
Comment (skabashnyuk) [inline wsmaster/codenvy-hosted-api-organization/src/main/java/com/codenvy/organization/api/OrganizationManager.java]: unmodifiable?

Comment (skabashnyuk) [inline wsmaster/codenvy-hosted-api-organization/src/main/java/com/codenvy/organization/api/OrganizationManager.java]: to lower case?

File: wsmaster/codenvy-hosted-api-organization/src/main/java/com/codenvy/organization/api/OrganizationLinksInjector.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import java.util.List;
+
+import static javax.ws.rs.core.MediaType.APPLICATION_JSON;
+
+/**
+ * Helps to inject {@link OrganizationService} related links.
+ *
+ * @author Sergii Leschenko
+ */
+public class OrganizationLinksInjector {
```
Comment (evoevodin) [inline wsmaster/codenvy-hosted-api-organization/src/main/java/com/codenvy/organization/api/OrganizationLinksInjector.java]: Does it have to be a singleton ?

File: wsmaster/codenvy-hosted-api-organization/src/main/java/com/codenvy/organization/api/OrganizationManager.java [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Facade for Organization related operations.
+ *
+ * @author gazarenkov
+ * @author Sergii Leschenko
+ */
+public class OrganizationManager {
```
Comment (evoevodin) [inline wsmaster/codenvy-hosted-api-organization/src/main/java/com/codenvy/organization/api/OrganizationManager.java]: Does it have to be a singleton?

File: wsmaster/codenvy-hosted-api-organization/src/main/java/com/codenvy/organization/api/OrganizationManager.java [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+    public Organization create(Organization newOrganization) throws ConflictException, ServerException {
+        requireNonNull(newOrganization, "Required non-null organization");
+        if (reservedNames.contains(newOrganization.getName().toLowerCase())) {
+            throw new ConflictException(String.format("Organization name '%s' is reserved", newOrganization.getName()));
+        }
+        final OrganizationImpl organization = new OrganizationImpl(NameGenerator.generate("organization", 16),
+                                                                   newOrganization.getName(),
+                                                                   newOrganization.getParent());
+        organizationDao.create(organization);
+        eventService.publish(new OrganizationCreatedEvent(organization));
```
Comment (evoevodin) [inline wsmaster/codenvy-hosted-api-organization/src/main/java/com/codenvy/organization/api/OrganizationManager.java]: Are you sure this event shouldn't be published in data layer?

File: wsmaster/codenvy-hosted-api-organization/src/main/java/com/codenvy/organization/api/OrganizationManager.java [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+     * @throws ServerException
+     *         when any other error occurs organization updating
+     */
+    public Organization update(String organizationId, Organization update) throws NotFoundException,
+                                                                                      ConflictException,
+                                                                                      ServerException {
+        requireNonNull(organizationId, "Required non-null organization id");
+        requireNonNull(update, "Required non-null organization");
+        if (reservedNames.contains(update.getName().toLowerCase())) {
+            throw new ConflictException(String.format("Organization name '%s' is reserved", update.getName()));
```
Comment (evoevodin) [inline wsmaster/codenvy-hosted-api-organization/src/main/java/com/codenvy/organization/api/OrganizationManager.java]: `create` and `update` methods share the same logic, it think this logic deserves a separate method, wdyt?

File: wsmaster/codenvy-hosted-api-organization/src/main/java/com/codenvy/organization/spi/jpa/JpaMemberDao.java [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+        final EntityManager manager = managerProvider.get();
+        try {
+            List<MemberImpl> members = manager.createNamedQuery("Member.getMember", MemberImpl.class)
+                                              .setParameter("userId", userId)
+                                              .setParameter("organizationId", organizationId)
+                                              .getResultList();
+            if (!members.isEmpty()) {
+                manager.remove(members.get(0));
+            }
+        } catch (NoResultException ignored) {
```
Comment (evoevodin) [inline wsmaster/codenvy-hosted-api-organization/src/main/java/com/codenvy/organization/spi/jpa/JpaMemberDao.java]: i don't see anyone who can throw this exception, do you?

Comment (sleshchenko) [inline wsmaster/codenvy-hosted-api-organization/src/main/java/com/codenvy/organization/spi/jpa/JpaMemberDao.java]: Neither do I. Thanks I'll remove it

File: wsmaster/codenvy-hosted-api-organization/src/main/java/com/codenvy/organization/api/OrganizationLinksInjector.java (hunk 2 above)
Comment (sleshchenko) [inline wsmaster/codenvy-hosted-api-organization/src/main/java/com/codenvy/organization/api/OrganizationLinksInjector.java]: Marked this class as singleton

File: wsmaster/codenvy-hosted-api-organization/src/main/java/com/codenvy/organization/api/OrganizationManager.java (hunk 3 above)
Comment (sleshchenko) [inline wsmaster/codenvy-hosted-api-organization/src/main/java/com/codenvy/organization/api/OrganizationManager.java]: Marked this class as singleton

File: wsmaster/codenvy-hosted-api-organization/src/main/java/com/codenvy/organization/api/OrganizationManager.java (hunk 4 above)
Comment (sleshchenko) [inline wsmaster/codenvy-hosted-api-organization/src/main/java/com/codenvy/organization/api/OrganizationManager.java]: OrganizationCreatedEvent renamed to OrganizationPersistedEvent and move publishing of it to OrganizationDao

File: wsmaster/codenvy-hosted-api-organization/src/main/java/com/codenvy/organization/api/OrganizationManager.java (hunk 5 above)
Comment (sleshchenko) [inline wsmaster/codenvy-hosted-api-organization/src/main/java/com/codenvy/organization/api/OrganizationManager.java]: :+1: I'll rework it

**Pull request merged**